controllers/apim/ratelimitpolicy_controller.go

```diff
@@ -85,15 +100,343 @@ func (r *RateLimitPolicyReconciler) Reconcile(eventCtx context.Context, req ctrl
 		}
 	}
 
-	specResult, specErr := r.reconcileSpec(ctx, &rlp)
-	logger.Info("spec reconcile done", "result", specResult, "error", specErr)
-	if specErr == nil && specResult.Requeue {
-		logger.Info("Reconciling not finished. Requeueing.")
-		return specResult, nil
+	if err := r.reconcileLimits(ctx, &rlp); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	for _, networkingRef := range rlp.Spec.NetworkingRef {
+		switch networkingRef.Type {
+		case apimv1alpha1.NetworkingRefType_HR:
+			logger.Info("HTTPRoute is not implemented yet") // TODO(rahulanand16nov)
+			return ctrl.Result{}, nil
+		case apimv1alpha1.NetworkingRefType_VS:
+			vsNamespacedName := client.ObjectKey{
+				Namespace: rlp.Namespace,
+				Name:      networkingRef.Name,
+			}
+
+			var vs istio.VirtualService
+			if err := r.Client().Get(ctx, vsNamespacedName, &vs); err != nil {
+				if apierrors.IsNotFound(err) {
+					logger.Info("no virtualservice found", "lookup name", vsNamespacedName.String())
+					return ctrl.Result{}, nil
+				}
+				logger.Error(err, "failed to get VirutalService")
+				return ctrl.Result{}, err
+			}
+
+			if err := r.reconcileWithVirtualService(ctx, &vs, &rlp); err != nil {
+				logger.Error(err, "failed to reconcile with virtualservice")
+				return ctrl.Result{}, err
+			}
+		}
 	}
 	return ctrl.Result{}, nil
 }
 
+func (r *RateLimitPolicyReconciler) reconcileWithVirtualService(ctx context.Context, vs *istio.VirtualService, rlp *apimv1alpha1.RateLimitPolicy) error {
+	logger := r.Logger()
+
+	_, ok := vs.Annotations[VSAnnotationRateLimitPolicy]
+	if !ok {
+		rlpNamespacedName := client.ObjectKeyFromObject(rlp)
+		vs.Annotations[VSAnnotationRateLimitPolicy] = rlpNamespacedName.String()
+		if err := r.Client().Update(ctx, vs); err != nil {
```

Comment:
There needs to be a good reason to update a resource that the controller is not owner of. 

What is the purpose of adding that annotation?

Reply:
Yes not sure this is needed now that we have changed the spec to have a targetRef. Currently the rate limit policy looks in the same namespace. So we should be able to put a watch on VirtualServices and HTTPRoutes and when one changes in a way we need to do something about, look up any RLP in same namespace and identify any that target the changed resource.

Reply:
It provided a mechanism to map to a particular RLP when VirtualService changes and is being watched. Considering what Craig said, this can be removed and I'll update the PR.

Reply:
Do we want to remove this as part of this PR?

Reply:
Removing this and watching VS/HR is a relatively big change. I'll add them in the next PR.